File: backend/models/tracking.py
from typing import Literal

from pydantic import BaseModel, Field


class Coordinate(BaseModel):
    x: float
    y: float
    z: float


class BallState(BaseModel):
    position: Coordinate
    is_occluded: bool


class PlayerState(BaseModel):
    id: Literal["player_bottom", "player_top"]
    position: Coordinate


class ProcFrameData(BaseModel):
    frame_index: int
    ball: BallState
    players: list[PlayerState] = Field(min_length=2, max_length=2)
    ball_speed_kmh: float = Field(ge=0)
    spin_rate_rpm: float = Field(ge=0)
    hitter: Literal["p1", "p2"] | None = None


# Alias used by legacy code paths
FrameData = ProcFrameData

File: frontend/src/types/tracking.ts
export interface Coordinate {
  x: number;
  y: number;
  z: number;
}

export interface PlayerState {
  id: 'player_bottom' | 'player_top' | string;
  position: Coordinate;
}

export interface BallState {
  position: Coordinate;
  is_occluded: boolean;
}

export interface ProcFrameData {
  frame_index: number;
  ball: BallState;
  players: PlayerState[];
  ball_speed_kmh: number;
  spin_rate_rpm?: number;
  hitter?: 'p1' | 'p2' | null;
  shot_type?: string;
}

export interface SequenceResponse {
  sequence: ProcFrameData[];
}

export type JobStatus =
  | 'queued'
  | 'processing'
  | 'completed'
  | 'failed'
  | 'calibration_required'
  | 'expired';

export interface JobStatusResponse {
  job_id: string;
  status: JobStatus;
  progress: number;
  stage: string | null;
  stage_progress: number;
  frames_total: number;
  frames_processed: number;
  error: string | null;
  calibration_failed: boolean;
}

export interface JobUploadResponse {
  job_id: string;
  status: 'queued';
  message: string;
}

/** True when VITE_API_URL is set (local dev or production HF backend). */
export function isServerProcessingEnabled(): boolean {
  const url = import.meta.env.VITE_API_URL as string | undefined;
  return Boolean(url && url.trim().length > 0);
}

export function getApiBaseUrl(): string {
  const url = (import.meta.env.VITE_API_URL as string | undefined)?.trim();
  if (!url) throw new Error('VITE_API_URL is not configured');
  return url.replace(/\/$/, '');
}
